flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,213 @@
+ * care of re-uploading changes from the previous {@link #persist(SequenceNumber)} call, if those
+ * changes haven't been yet {@link #confirm(SequenceNumber, SequenceNumber)}'ed. This is crucial as
+ * until changes are {@link #confirm(SequenceNumber, SequenceNumber)}, they still can be aborted and
+ * removed/deleted.
+ *
+ * <p>For example if checkpoint N-1 fails and is disposed, after checkpoint N has already started.
+ * In this case, when we are persisting {@link StateChangeSet}s for checkpoint N, we need to
+ * re-upload {@link StateChangeSet}s that belonged to checkpoint N-1.
+ */
+@NotThreadSafe
+class FsStateChangelogWriter implements StateChangelogWriter<StateChangelogHandleStreamImpl> {
+    private static final Logger LOG = LoggerFactory.getLogger(FsStateChangelogWriter.class);
+
+    private final UUID logId;
+    private final KeyGroupRange keyGroupRange;
+    private final StateChangeStore store;
+    private final NavigableMap<SequenceNumber, StateChangeSet> changeSets = new TreeMap<>();
+    private final NavigableMap<SequenceNumber, StoreResult> uploaded = new TreeMap<>();
+    private final NavigableMap<SequenceNumber, StoreResult> confirmed = new TreeMap<>();
+    private List<StateChange> activeChangeSet = new ArrayList<>();
+    private SequenceNumber lastAppendedSequenceNumber = SequenceNumber.of(0L);
+    private boolean closed;
+    private final ChangelogCallbackExecutor executor;
+
+    FsStateChangelogWriter(
+            UUID logId,
+            KeyGroupRange keyGroupRange,
+            StateChangeStore store,
+            ChangelogCallbackExecutor executor) {
+        this.logId = logId;
+        this.keyGroupRange = keyGroupRange;
+        this.store = store;
+        this.executor = checkNotNull(executor);
+    }
+
+    @Override
+    public void append(int keyGroup, byte[] value) {
+        LOG.trace("append to {}: keyGroup={} {} bytes", logId, keyGroup, value.length);
+        checkState(!closed, "%s is closed", logId);
+        activeChangeSet.add(new StateChange(keyGroup, value));
+        // size threshold could be added to call persist when reached. considerations:
+        // 0. can actually degrade performance by amplifying number of requests
+        // 1. which range to persist?
+        // 2. how to deal with retries/aborts?
+    }
+
+    @Override
+    public SequenceNumber lastAppendedSequenceNumber() {
+        rollover();
+        LOG.trace("query {} sqn: {}", logId, lastAppendedSequenceNumber);
+        return lastAppendedSequenceNumber;
+    }
+
+    @Override
+    public CompletableFuture<StateChangelogHandleStreamImpl> persist(SequenceNumber from)
+            throws IOException {
+        LOG.debug("persist {} from {}", logId, from);
+        checkNotNull(from);
+        // todo: check range
+
+        rollover();
+        Collection<StoreResult> readyToReturn = confirmed.tailMap(from, true).values();
+        Collection<StateChangeSet> toUpload = changeSets.tailMap(from, true).values();
```

Comment:
I'm thinking how could we later implement pre-emptive upload (i.e. call `persist()` from `append()`).

In this version (#15371 or #15322 as opposed to #14839), we need to filter out already uploaded changes (`uploaded.subMap`) from `toUpload`.

But we also don't want to include the changes from the previous, not yet confirmed/nor aborted checkpoint. Which will also reside in `uploaded`. Do you have any idea how to implement this?

Reply:
So for pre-emptive uploads there will be a separate collection which will also be updated by a separate callback. We'll also have to truncate it. 
Sounds a bit complex but I think it should work.